src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3621,13 +3621,61 @@ private TypeWithState InferResultNullability(BinaryOperatorKind operatorKind, Me
         protected override void VisitBinaryOperatorChildren(ArrayBuilder<BoundBinaryOperator> stack)
         {
             var binary = stack.Pop();
-
             var (leftOperand, leftConversion) = RemoveConversion(binary.Left, includeExplicitConversions: false);
-            Visit(leftOperand);
+
+            // Only the leftmost operator of a left-associative binary operator chain can learn from a conditional access on the left
+            // For simplicity, we just special case it here.
+            // For example, `a?.b(out x) == true` has a conditional access on the left of the operator,
+            // but `expr == a?.b(out x) == true` has a conditional access on the right of the operator
+            if (VisitPossibleConditionalAccess(leftOperand, out var conditionalStateWhenNotNull)
+                && CanPropagateStateWhenNotNull(leftConversion)
+                && isEqualsOrNotEquals(binary))
+            {
+                Debug.Assert(!IsConditionalState);
+                var leftType = ResultType;
+                var (rightOperand, rightConversion) = RemoveConversion(binary.Right, includeExplicitConversions: false);
+
+                // Consider the following two scenarios:
+                // `a?.b(x = new object()) == c.d(x = null)` // `x` is maybe-null after expression
+                // `a?.b(x = null) == c.d(x = new object())` // `x` is not-null after expression
+
+                // In order to properly integrate state changes from the RHS into the LHS's "state when not null", we can't simply Meet or Join the state.
+                // We have to visit it twice. Once without diagnostics using the LHS's "state when not null", and again with the "state when maybe null" for diagnostics and public API.
+                var oldDisableDiagnostics = _disableDiagnostics;
```

Comment:
This part is fairly gnarly. I wasn't able to find a complete solution that didn't involve double-visiting the right operand. Because the operator chain is left-associative, we think that re-visiting the right in this scenario is not that expensive.

If this is found to be too ugly of a solution, we could also consider adding a requirement that the "other" operand is a constant value in order to learn from a conditional access.